syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt

```diff
@@ -83,49 +90,50 @@ class BlockPuller internal constructor(private val connectionHandler: Connection
-                    connectionHandler.sendMessage(Request.newBuilder()
-                            .setId(requestId)
-                            .setFolder(fileBlocks.folder)
-                            .setName(fileBlocks.path)
-                            .setOffset(block.offset)
-                            .setSize(block.size)
-                            .setHash(ByteString.copyFrom(Hex.decode(block.hash)))
-                            .build())
-                    logger.debug("sent request for block, hash = {}", block.hash)
+
+        for (block in fileBlocks.blocks.distinctBy { it.hash }) {
+            val requestId = responseHandler.registerListener {
+                response ->
+
+                try {
+                    NetworkUtils.assertProtocol(response.code == ErrorCode.NO_ERROR, { "received error response, code = ${response.code}" })
+                    val data = response.data.toByteArray()
+                    val hash = Hex.toHexString(MessageDigest.getInstance("SHA-256").digest(data))
+
+                    if (hash != block.hash) {
+                        throw IllegalStateException("expected block with hash ${block.hash}, but got block with hash $hash")
+                    }
+
+                    synchronized(lock) {
+                        blocksByHash[hash] = data
+                        logger.debug("aquired block, hash = {}", hash)
+                        lock.notify()
+                    }
+                } catch (ex: Exception) {
+                    error.compareAndSet(null, ex)
+
+                    synchronized(lock) {
+                        lock.notify()
+                    }
                 }
             }
-            return fileDownloadObserver
-        }
-    }
 
-    fun onResponseMessageReceived(response: BlockExchangeProtos.Response) {
-        synchronized(lock) {
-            if (!requestIds.contains(response.id)) {
-                return
-            }
-            NetworkUtils.assertProtocol(response.code == ErrorCode.NO_ERROR, {"received error response, code = ${response.code}"})
-            val data = response.data.toByteArray()
-            val hash = Hex.toHexString(MessageDigest.getInstance("SHA-256").digest(data))
-            if (missingHashes.remove(hash)) {
-                blocksByHash.put(hash, data)
-                logger.debug("aquired block, hash = {}", hash)
-                lock.notify()
-            } else {
-                logger.warn("received not-needed block, hash = {}", hash)
-            }
+            connectionHandler.sendMessage(Request.newBuilder()
+                    .setId(requestId)
+                    .setFolder(fileBlocks.folder)
+                    .setName(fileBlocks.path)
+                    .setOffset(block.offset)
+                    .setSize(block.size)
+                    .setHash(ByteString.copyFrom(Hex.decode(block.hash)))
+                    .build())
+
+            logger.debug("sent request for block, hash = {}", block.hash)
```

Comment:
You should probably queue up atleast 4mb worth of requests in a single go, otherwise it will be terribly slow for 128kb blocks.

Reply:
All blocks are requested at the same time (which makes canceling not possible) currently so it should not increase the delay (I assume that there are no network packages with 4 MB and that the server handles small blocks well or has got an SSD).

I'm still going to change the block size. I think that later, 4 active block requests per time would fit (should be enough to make sure that there is always something transmitted + canceling should take effect without big delay).

Reply:
I prefer a block size of not more than 512 KiB instead of 4 MiB because 4 MiB data transfer if the user canceled anddownload is too much.

Reply:
The block size is decided at scan time, so you have to ask exactly the amount as specified in the block object, and you cannot change that, but you should limit not to have more than 4mb of requests in flight at any point in time.

Reply:
@AudriusButkevicius The receiver can chose any range of a file for a request.

https://docs.syncthing.net/specs/bep-v1.html#selection-of-block-size

> An implementation MUST therefore accept files with a block size differing from the above rule. This does not mean that arbitrary block sizes are allowed. The block size used MUST be exactly one of the power-of-two block sizes listed in the table above.

https://docs.syncthing.net/specs/bep-v1.html#request

> The folder and name fields are as documented for the Index message. The offset and size fields specify the region of the file to be transferred. This SHOULD equate to exactly one block as seen in an Index message.

Reply:
The implementation can choose to the block size of power of two at scan time, but once the block size is decided you have to ask for exactly that, as asking for more data means the hash will not match.

Reply:
The implementation can choose to the block size of power of two at scan time, but once the block size is decided you have to ask for exactly that, as asking for more data means the hash will not match.

Reply:
@AudriusButkevicius Where is this documented?